crates/build/src/build.rs

```diff
@@ -206,3 +211,132 @@ fn print_elf_paths_cargo_directives(target_elf_paths: &[(String, Utf8PathBuf)])
+    }
+
+    // We need an exception for our test fixtures.
+    if std::env::var("CARGO_PKG_NAME").unwrap_or_default() == "test-artifacts" {
+        return Ok(());
+    }
+
+    // This might be a workspace, need to optionally search parent dirs for lock files.
+    let canon =
+        program_dir.as_ref().canonicalize().expect("Failed to canonicalize path to program dir");
+
+    let mut lock_path = canon.join("Cargo.lock");
+    if !lock_path.is_file() {
+        let mut curr_path: &Path = canon.as_ref();
+
+        while let Some(parent) = curr_path.parent() {
+            let maybe_lock_path = parent.join("Cargo.lock");
+
+            if maybe_lock_path.is_file() {
+                lock_path = maybe_lock_path;
+                break;
+            } else {
+                curr_path = parent;
+            }
+        }
+
+        if !lock_path.is_file() {
+            panic!("Failed to find Cargo.lock in the program directory or its parent directories");
+        }
+    }
+
+    println!("cargo:warning=Found Cargo.lock at {}", lock_path.display());
+
+    let lock_file = cargo_lock::Lockfile::load(lock_path).expect("Failed to load Cargo.lock");
+
+    // You need an entrypoint, therefore you need sp1-zkvm.
+    let vm_version = &lock_file
+        .packages
+        .iter()
+        .find(|p| p.name.as_str() == "sp1-zkvm")
+        .expect("The guest program should have a sp1-zkvm dependency")
+        .version;
+
+    let maybe_sdk =
+        lock_file.packages.iter().find(|p| p.name.as_str() == "sp1-sdk").map(|p| &p.version);
+
+    // Print these just to be useful.
+    // Most people will install the actual rust toolchain along with their cargo-prove,
+    // so we can just use the cargo-prove version.
+    let toolchain_version = {
+        let output = std::process::Command::new("cargo")
+            .args(["prove", "--version"])
+            .output()
+            .expect("Failed to find check toolchain version, see https://docs.succinct.xyz/getting-started/install.html");
+
+        let version = String::from_utf8(output.stdout).expect("Failed to parse version string");
+
+        semver::Version::parse(
+            version
+                .split_once('\n')
+                .expect("The version should have whitespace, this is a bug")
+                .1
+                .trim(),
+        )
+        .expect("Failed to parse version string, this is a bug")
+    };
+
+    if verbose {
+        maybe_sdk.inspect(|v| {
+            println!("cargo:warning=Locked version of sp1-sdk is {}", v);
+        });
+        println!("cargo:warning=Locked version of sp1-zkvm is {}", vm_version);
+        println!("cargo:warning=Current toolchain version = {}", toolchain_version);
+    }
+
+    if vm_version.major != toolchain_version.major || vm_version.minor != toolchain_version.minor {
+        return Err(anyhow::anyhow!(
+            "Locked version of sp1-zkvm is incompatible with the current toolchain version"
+        ));
+    }
```

Comment:
We should perform the following check here:

- `cargo prove` should always be backwards-compatible within the same major version. Ex. `cargo prove` `1.2` is compat with `sp1-zkvm` `1.1`. However, `sp1-zkvm` `1.2` is not guaranteed to be compat with `cargo prove` `1.1`.